modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/service/ClientServiceInvokeRequest.java

```diff
@@ -0,0 +1,324 @@
+            grp = (ClusterGroupAdapter)grp.forSubjectId(ctx.securityContext().subject().id());
+
+        grp = (ClusterGroupAdapter)(nodeIds.isEmpty() ? grp.forServers() : grp.forNodeIds(nodeIds));
+
+        IgniteServices services = grp.services();
+
+        if (!keepBinary() && args.length > 0) {
+            for (int i = 0; i < args.length; i++) {
+                if (paramTypeIds != null)
+                    reader.readInt(); // Skip parameter typeId, we already read it in constructor.
+
+                args[i] = reader.readObject();
+            }
+        }
+
+        try {
+            Object res;
+
+            if (PlatformService.class.isAssignableFrom(svcCls)) {
+                PlatformService proxy = services.serviceProxy(name, PlatformService.class, false, timeout);
+
+                res = proxy.invokeMethod(methodName, keepBinary(), !keepBinary(), args);
+            }
+            else {
+                GridServiceProxy<?> proxy = new GridServiceProxy<>(grp, name, Service.class, false, timeout,
+                    ctx.kernalContext());
+
+                Method method = resolveMethod(ctx, svcCls);
+
+                res = proxy.invokeMethod(method, args);
+            }
+
+            return new ClientObjectResponse(requestId(), res);
+        }
+        catch (PlatformNativeException e) {
+            ctx.kernalContext().log(getClass()).error("Failed to invoke platform service", e);
+
+            throw new IgniteException("Failed to invoke platform service, see server logs for details");
+        }
+        catch (Throwable e) {
+            throw new IgniteException(e);
+        }
+    }
+
+    /**
+     * Keep binary flag.
+     */
+    private boolean keepBinary() {
+        return (flags & FLAG_KEEP_BINARY_MASK) != 0;
+    }
+
+    /**
+     * "Has parameter types" flag.
+     */
+    private boolean hasParameterTypes() {
+        return (flags & FLAG_PARAMETER_TYPES_MASK) != 0;
+    }
+
+    /**
+     * @param ctx Connection context.
+     * @param name Service name.
+     */
+    private static ServiceDescriptor findServiceDescriptor(ClientConnectionContext ctx, String name) {
+        for (ServiceDescriptor desc : ctx.kernalContext().service().serviceDescriptors()) {
+            if (name.equals(desc.name()))
+                return desc;
+        }
+
+        throw new IgniteException("Service not found: " + name);
+    }
+
+    /**
+     * Resolve method by method name and parameter types or parameter values.
+     */
+    private Method resolveMethod(ClientConnectionContext ctx, Class<?> cls) throws ReflectiveOperationException {
+        if (paramTypeIds != null) {
+            MethodDescriptor desc = new MethodDescriptor(cls, methodName, paramTypeIds);
+
+            Method method = methodsCache.get(desc);
+
+            if (method != null)
+                return method;
+
+            BinaryContext bctx = ((CacheObjectBinaryProcessorImpl)ctx.kernalContext().cacheObjects()).binaryContext();
```

Comment:
Let's use public API and avoid type casts to internal classes when possible. This context is only needed for `ctx.typeId` call, which is available in `IgniteBinary`.

Reply:
Fixed